[PyTorch Numeric Suite] Add support for dynamic quantization of linear module

hx89 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40065 [PyTorch Numeric Suite] Add support for dynamic LSTM
* **#39024 [PyTorch Numeric Suite] Add support for dynamic quantization of linear module**

Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

Commit:
[PyTorch Numeric Suite] Add support for dynamic quantization of linear module

Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -89,6 +90,10 @@ def add_observer_(module):
         if type(child) == nnq.FloatFunctional or type(child) == nnq.QFunctional:
             if hasattr(child, 'qconfig') and child.qconfig is not None:
                 child.activation_post_process = child.qconfig.activation()
+        elif type(child) == nnqd.Linear or type(child) == nnq.Linear:
```
Why do you need to special case nnq.Linear?

hx89 replied:
It's because nnq.Linear has child module _packed_params and existing code won't assign qconfig to it.

raghuramank100 replied:
You can do this instead:
``` python
_INCLUDE_NUMERIC_SUITE_PROPAGATE_LIST = {
    nnq.Linear._packed_params
}
DEFAULT_NUMERIC_SUITE_COMPARE_MODEL_OUTPUT_WHITE_LIST = (
    set(DEFAULT_MODULE_MAPPING.values())
    | set(DEFAULT_QAT_MODULE_MAPPING.values())
    | set(DEFAULT_DYNAMIC_MODULE_MAPPING.values())
    | set(DEFAULT_MODULE_MAPPING.keys())
    | set(DEFAULT_QAT_MODULE_MAPPING.keys())
    | set(DEFAULT_DYNAMIC_MODULE_MAPPING.keys())
    | _INCLUDE_QCONFIG_PROPAGATE_LIST
    |_INCLUDE_NUMERIC_SUITE_PROPAGATE_LIST
) - _EXCLUDE_QCONFIG_PROPAGATE_LIST
```

Also, can you move the list definitions here to _numeric_suite.py so that default_mappings only has the code specific to quantize/prepare/convert?

hx89 replied:
My previous comment is not correct, the issue should be nnq.Linear has child module _packed_params and add_observer() only add observer to leaf node, so adding nnq.Linear._packed_params to white list may not help in this case.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -51,6 +64,15 @@ def compare_weights(float_dict, quantized_dict):
             weight_dict[key]["quantized"] = quantized_dict[key]
+            continue
+
+        # For matching "fc.weight" and "fc._packed_params._packed_params"
```
You will need to look at the version of the state dict to figure out what needs to be done here. State dicts will change over time, so having  a check for the version number is a good idea. This is available at the level of each module's state dict.

hx89 replied:
Could you give an example how to check the state dict version?

raghuramank100 replied:
The state dict has a field called _version for each module, an example is at:
https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/linear.py#L11
The main code in module that does state dict serialization is at: 
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L851

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic quantization of linear module"


Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic quantization of linear module"


Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

[ghstack-poisoned]

raghuramank100 reviewed (changes requested):
Some suggestions

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic quantization of linear module"


Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic quantization of linear module"


Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -73,7 +73,7 @@ def _observer_forward_hook(self, input, output):
     return self.activation_post_process(output)
 
-def add_observer_(module, device=None):
+def add_observer_(module, white_list=None, device=None):
```
Please use supported_list instead of white_list

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -142,7 +147,7 @@ def add_quant_dequant(module):
     return module
 
-def prepare(model, inplace=False, white_list=DEFAULT_QCONFIG_PROPAGATE_WHITE_LIST):
+def prepare(model, inplace=False, white_list=DEFAULT_QCONFIG_PROPAGATE_WHITE_LIST, observer_white_list=None):
```
Similar comment, 
Also, please update the docstrings

raghuramank100 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -252,9 +252,9 @@ def forward(self, x):
 
 class SingleLayerLinearDynamicModel(torch.nn.Module):
-    def __init__(self):
+    def __init__(self, qengine='fbgemm'):
```
We normally use @override_qengines for sweeping over qengines. Please do not pass qengine as a parameter to the model, instead in the test, specify the qengine using the @override_qengines

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -108,6 +113,33 @@ def compare_and_validate_results(float_model, q_model):
+                    compare_and_validate_results(model, q_model)
+
+        # Test dynamic quantization
+        for qengine in supported_qengines:
```
Use @override_qengines instead of looping here

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -95,8 +98,10 @@ def compare_and_validate_results(float_model, q_model):
             for k, v in weight_dict.items():
                 self.assertTrue(v["float"].shape == v["quantized"].shape)
 
+        # Test static quantization
```
This test is too big now, can you refactor into smaller tests: Perhaps one for Conv, one for Linear and one for dynamic quant?

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -135,6 +168,24 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
                     q_model = quantize(model, default_eval_fn, self.img_data)
                     compare_and_validate_results(model, q_model, module_swap_list, data)
 
+                # Test Linear
```
Similar comment here, please refactor test_compare_model_stub into smaller units

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -189,14 +257,30 @@ def compare_and_validate_results(float_model, q_model, data):
                     q_model = quantize(model, default_eval_fn, self.img_data)
                     compare_and_validate_results(model, q_model, data)
 
+                # Test Linear
```
Suggest in general breaking each of these tests into 4 smaller tests: Conv/ConvBN, Linear, DynamicLinear and Functional.

raghuramank100 reviewed (changes requested):
Looks good, a few comments

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic quantization of linear module"


Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic quantization of linear module"


Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic quantization of linear module"


Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

[ghstack-poisoned]

hx89 commented:
Address comments.

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic quantization of linear module"


Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -82,6 +82,7 @@ def add_observer_(module, device=None):
     Args:
         module: input module with qconfig attributes for all the leaf modules that we want to quantize
         device: parent device, if any
+        supported_list: list of non-leaf modules we want to add observer
```
Lets call this non_leaf_module_list instead to that it is clear

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -142,7 +147,7 @@ def add_quant_dequant(module):
     return module
 
-def prepare(model, inplace=False, white_list=DEFAULT_QCONFIG_PROPAGATE_WHITE_LIST):
+def prepare(model, inplace=False, white_list=DEFAULT_QCONFIG_PROPAGATE_WHITE_LIST, observer_supported_list=None):
```
Similar comment, lets call this: observer_non_leaf_module_list

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic quantization of linear module"


Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic quantization of linear module"


Add support for dynamic quantization of linear module.

Differential Revision: [D21675971](https://our.internmc.facebook.com/intern/diff/D21675971/)

[ghstack-poisoned]
